### PR TITLE
Add more table helpers

### PR DESCRIPTION
**Pull request by rubenwardy**
This is useful for the following code

```lua
print("there are " .. things:count() .. " things!")

local sorted_names = things:map(function(e)
    return { name = e.name, score = calcScore(e) }
end):sort(function(a, b)
    return a.score > b.score
end)
```

### CODE REVIEW

File: builtin/common/misc_helpers.lua [hunk 1]
```diff
@@ -198,9 +198,44 @@ function table.indexof(list, val)
...
+	local i = 0
+	for _ in pairs(t) do
+		i = i + 1
+	end
+	return i
+end
+
+--------------------------------------------------------------------------------
+function table.random(t)
+	return t[math.random(#t)]
```
Comment (sfan5) [inline builtin/common/misc_helpers.lua]: This doesn't work with tables that are not lists.
This restriction is not documented either.

File: builtin/common/misc_helpers.lua [hunk 2]
```diff
@@ -198,9 +198,44 @@ function table.indexof(list, val)
...
 assert(table.indexof({"foo", "bar"}, "foo") == 1)
 assert(table.indexof({"foo", "bar"}, "baz") == -1)
 
+--------------------------------------------------------------------------------
+function table.count(t)
+	local i = 0
+	for _ in pairs(t) do
+		i = i + 1
+	end
+	return i
```
Comment (sfan5) [inline builtin/common/misc_helpers.lua]: What's the usecase here?

File: builtin/common/misc_helpers.lua [hunk 3]
```diff
@@ -198,9 +198,44 @@ function table.indexof(list, val)
...
+		ret[key] = f(value)
+	end
+	return ret
+end
+
+--------------------------------------------------------------------------------
+function table.map_inplace(t, f)
+	for key, value in pairs(t) do
+		value = f(value)
+		if value ~= nil then
```
Comment (sfan5) [inline builtin/common/misc_helpers.lua]: this behaviour does not match what `.map` does, it's not documented either

Comment (rubenwardy) [reply builtin/common/misc_helpers.lua]: that was a mistake

File: builtin/common/misc_helpers.lua (hunk 1 above)
Comment (rubenwardy) [reply builtin/common/misc_helpers.lua]: fixed

File: doc/lua_api.txt [hunk 4]
```diff
@@ -2491,6 +2491,42 @@ For the following functions `x` can be either a vector or a number:
...
+    * `f` - function(element), return new value
+    * Each element of `t` is passed to `f` and replaced with the returned value.
+    * `t` **is** modified.
+
+### String helpers
+
+* `string.split(str, separator, include_empty, max_splits, sep_is_pattern)`
+    * `separator`: string, default: `","`
+    * `include_empty`: boolean, default: `false`
+    * `max_splits`: number, if it's positive, splits aren't limited,
```
Comment (SmallJoker) [inline doc/lua_api.txt]: This was only moved but the description seems to be wrong:
`if it's positive, splits are limited,`

**Pull request closed without merging**